File: frontend/lib/work.ts
import { apiFetch } from './api';
import type { AgentType } from './agents';

export type TaskStatus =
  | 'todo'
  | 'in_progress'
  | 'blocked'
  | 'waiting_for_user'
  | 'done'
  | 'archived';

export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent';
export type ArtifactFormat = 'markdown' | 'csv' | 'pdf' | 'text';
export type DigestCadence = 'daily' | 'weekly';
export type MonitorType = 'market' | 'competitor' | 'follow_up' | 'risk';
export type MonitorStatus = 'active' | 'paused';

export interface Task {
  id: string;
  project_id: string;
  source_run_id: string | null;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  owner_agent_type: AgentType | null;
  due_at: string | null;
  blocked_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskCreatePayload {
  title: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  owner_agent_type?: AgentType | null;
  due_at?: string | null;
  blocked_reason?: string | null;
}

export type TaskUpdatePayload = Partial<TaskCreatePayload>;

export interface TaskEvent {
  id: string;
  task_id: string;
  project_id: string;
  actor_type: 'founder' | 'agent' | 'cofounder' | 'system';
  actor_label: string | null;
  event_type: string;
  summary: string;
  metadata_json: Record<string, unknown>;
  created_at: string;
}

export interface AgentRunStep {
  id: string;
  run_id: string;
  project_id: string;
  agent_type: AgentType;
  sequence: number;
  event_type: string;
  title: string;
  detail: string | null;
  payload: Record<string, unknown>;
  created_at: string;
}

export interface ArtifactVersion {
  id: string;
  artifact_id: string;
  version: number;
  content: string;
  metadata_json: Record<string, unknown>;
  created_by: string;
  created_at: string;
}

export interface Artifact {
  id: string;
  project_id: string;
  task_id: string | null;
  run_id: string | null;
  title: string;
  artifact_type: string;
  format: ArtifactFormat;
  current_version_id: string | null;
  created_at: string;
  updated_at: string;
  current_version: ArtifactVersion | null;
}

export interface ArtifactCreatePayload {
  title: string;
  artifact_type:
    | 'competitor_analysis'
    | 'pricing_model'
    | 'investor_update'
    | 'landing_page_copy'
    | 'roadmap'
    | 'general';
  format: ArtifactFormat;
  content: string;
  task_id?: string | null;
  metadata_json?: Record<string, unknown>;
}

export interface WorkQueue {
  today: Task[];
  blocked: Task[];
  waiting_for_you: Task[];
  upcoming: Task[];
  recent_artifacts: Artifact[];
  recent_steps: AgentRunStep[];
  unresolved_clarifications: number;
  stale_task_count: number;
}

export interface CofounderDigest {
  id: string;
  project_id: string;
  cadence: DigestCadence;
  title: string;
  summary: string;
  payload: Record<string, unknown>;
  created_at: string;
}

export interface CofounderMonitor {
  id: string;
  project_id: string;
  title: string;
  monitor_type: MonitorType;
  query: string;
  cadence: DigestCadence;
  status: MonitorStatus;
  last_checked_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface MonitorCreatePayload {
  title: string;
  monitor_type: MonitorType;
  query: string;
  cadence: DigestCadence;
}

export async function getWorkQueue(projectId: string): Promise<WorkQueue> {
  return apiFetch<WorkQueue>(`/projects/${projectId}/work-queue`);
}

export async function listTasks(projectId: string): Promise<Task[]> {
  return apiFetch<Task[]>(`/projects/${projectId}/tasks`);
}

export async function createTask(
  projectId: string,
  payload: TaskCreatePayload
): Promise<Task> {
  return apiFetch<Task>(`/projects/${projectId}/tasks`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function updateTask(
  projectId: string,
  taskId: string,
  payload: TaskUpdatePayload
): Promise<Task> {
  return apiFetch<Task>(`/projects/${projectId}/tasks/${taskId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
}

export async function listArtifacts(projectId: string): Promise<Artifact[]> {
  return apiFetch<Artifact[]>(`/projects/${projectId}/artifacts`);
}

export async function createArtifact(
  projectId: string,
  payload: ArtifactCreatePayload
): Promise<Artifact> {
  return apiFetch<Artifact>(`/projects/${projectId}/artifacts`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function listDigests(projectId: string): Promise<CofounderDigest[]> {
  return apiFetch<CofounderDigest[]>(`/projects/${projectId}/cofounder/digests`);
}

export async function createDigest(
  projectId: string,
  cadence: DigestCadence
): Promise<CofounderDigest> {
  return apiFetch<CofounderDigest>(`/projects/${projectId}/cofounder/digests`, {
    method: 'POST',
    body: JSON.stringify({ cadence }),
  });
}

export async function listMonitors(projectId: string): Promise<CofounderMonitor[]> {
  return apiFetch<CofounderMonitor[]>(`/projects/${projectId}/cofounder/monitors`);
}

export async function createMonitor(
  projectId: string,
  payload: MonitorCreatePayload
): Promise<CofounderMonitor> {
  return apiFetch<CofounderMonitor>(`/projects/${projectId}/cofounder/monitors`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

File: backend/app/schemas/work.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.agents import AgentType


TaskStatus = Literal[
    "todo",
    "in_progress",
    "blocked",
    "waiting_for_user",
    "done",
    "archived",
]
TaskPriority = Literal["low", "medium", "high", "urgent"]
ActorType = Literal["founder", "agent", "cofounder", "system"]
ArtifactFormat = Literal["markdown", "csv", "pdf", "text"]
ArtifactType = Literal[
    "competitor_analysis",
    "pricing_model",
    "investor_update",
    "landing_page_copy",
    "roadmap",
    "general",
]
DigestCadence = Literal["daily", "weekly"]
MonitorType = Literal["market", "competitor", "follow_up", "risk"]
MonitorStatus = Literal["active", "paused"]


class TaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)
    status: TaskStatus = "todo"
    priority: TaskPriority = "medium"
    owner_agent_type: AgentType | None = None
    due_at: datetime | None = None
    blocked_reason: str | None = Field(default=None, max_length=2000)


class TaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)
    status: TaskStatus | None = None
    priority: TaskPriority | None = None
    owner_agent_type: AgentType | None = None
    due_at: datetime | None = None
    blocked_reason: str | None = Field(default=None, max_length=2000)


class TaskResponse(BaseModel):
    id: str
    project_id: str
    source_run_id: str | None
    title: str
    description: str | None
    status: TaskStatus
    priority: TaskPriority
    owner_agent_type: AgentType | None
    due_at: datetime | None
    blocked_reason: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TaskEventCreate(BaseModel):
    event_type: str = Field(min_length=1, max_length=80)
    summary: str = Field(min_length=1, max_length=4000)
    actor_type: ActorType = "founder"
    actor_label: str | None = Field(default=None, max_length=120)
    metadata_json: dict = Field(default_factory=dict)


class TaskEventResponse(BaseModel):
    id: str
    task_id: str
    project_id: str
    actor_type: ActorType
    actor_label: str | None
    event_type: str
    summary: str
    metadata_json: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class AgentRunStepResponse(BaseModel):
    id: str
    run_id: str
    project_id: str
    agent_type: AgentType
    sequence: int
    event_type: str
    title: str
    detail: str | None
    payload: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class ArtifactCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    artifact_type: ArtifactType = "general"
    format: ArtifactFormat = "markdown"
    content: str = Field(min_length=1, max_length=100000)
    task_id: str | None = None
    metadata_json: dict = Field(default_factory=dict)


class ArtifactVersionCreate(BaseModel):
    content: str = Field(min_length=1, max_length=100000)
    created_by: str = Field(default="founder", max_length=120)
    metadata_json: dict = Field(default_factory=dict)


class ArtifactVersionResponse(BaseModel):
    id: str
    artifact_id: str
    version: int
    content: str
    metadata_json: dict
    created_by: str
    created_at: datetime

    model_config = {"from_attributes": True}


class ArtifactResponse(BaseModel):
    id: str
    project_id: str
    task_id: str | None
    run_id: str | None
    title: str
    artifact_type: str
    format: ArtifactFormat
    current_version_id: str | None
    created_at: datetime
    updated_at: datetime
    current_version: ArtifactVersionResponse | None = None

    model_config = {"from_attributes": True}


class WorkQueueResponse(BaseModel):
    today: list[TaskResponse]
    blocked: list[TaskResponse]
    waiting_for_you: list[TaskResponse]
    upcoming: list[TaskResponse]
    recent_artifacts: list[ArtifactResponse]
    recent_steps: list[AgentRunStepResponse]
    unresolved_clarifications: int
    stale_task_count: int


class CofounderDigestCreate(BaseModel):
    cadence: DigestCadence = "daily"


class CofounderDigestResponse(BaseModel):
    id: str
    project_id: str
    cadence: DigestCadence
    title: str
    summary: str
    payload: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class CofounderMonitorCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    monitor_type: MonitorType
    query: str = Field(min_length=1, max_length=4000)
    cadence: DigestCadence = "weekly"


class CofounderMonitorUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    monitor_type: MonitorType | None = None
    query: str | None = Field(default=None, min_length=1, max_length=4000)
    cadence: DigestCadence | None = None
    status: MonitorStatus | None = None
    last_checked_at: datetime | None = None


class CofounderMonitorResponse(BaseModel):
    id: str
    project_id: str
    title: str
    monitor_type: MonitorType
    query: str
    cadence: DigestCadence
    status: MonitorStatus
    last_checked_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
